FSI/Displacement.C

```diff
@@ -41,9 +41,43 @@ void preciceAdapter::FSI::Displacement::write(double* buffer, bool meshConnectiv
     * the outer for the locations and the inner for the dimensions.
     * See the preCICE writeBlockVectorData() implementation.
     */
-    FatalErrorInFunction
-        << "Writing displacements is not supported."
-        << exit(FatalError);
+
+    // Copy the displacement field from OpenFOAM to the buffer
+
+    if (this->locationType_ == LocationType::faceCenters)
+    {
+        // For every boundary patch of the interface
+        for (const label patchID : patchIDs_)
+        {
+            // Write the displacement to the preCICE buffer
+            // For every cell of the patch
+            forAll(cellDisplacement_->boundaryField()[patchID], i)
+            {
+                for (unsigned int d = 0; d < dim; ++d)
+                    buffer[i * dim + d] =
+                        cellDisplacement_->boundaryField()[patchID][i][d];
+            }
+        }
+    }
+    else if (this->locationType_ == LocationType::faceNodes)
+    {
+        // For every boundary patch of the interface
+        for (const label patchID : patchIDs_)
+        {
+            // Write the displacement to the preCICE buffer
+            // For every cell of the patch
+            forAll(pointDisplacement_->boundaryField()[patchID], i)
+            {
+                const labelList& meshPoints =
+                    mesh_.boundaryMesh()[patchID].meshPoints();
+
+                for (unsigned int d = 0; d < dim; ++d)
+                    buffer[i * dim + d] =
+                        //displacementField.boundaryField()[patchID][i][d];
+                        pointDisplacement_->internalField()[meshPoints[i]][d];
```

Comment:
pointVectorFields (and point*Fields in general) do not store any values on the patch itself; instead, all point values (internal points and points on the boundary) are stored in the internalField. So to set boundary point values, we need to use the meshPoints addressing to address into the internalField.

Reply:
But this is the `write()` function (which extracts values from the boundary and gives them to preCICE / the other side). Isn't `boundaryField` already interpolating on the boundary? Otherwise, if we give the values from the internalField, we may have a half-cell difference. In all other readers, we extract the values either from the face nodes, or from the face centers, but not from the cell centers*.

I was mainly puzzled since the loop at line 71 goes over boundaryField, while we give the internalField to the buffer.

*: That will be the case in volume coupling, which is a different story.

Reply:
For GeometryFields in OpenFOAM, data is stored as follows:

- volFields, e.g. volScalarField
    * internalField: for all cell-centre values
    * boundaryField: this is a list of fields (one for each patch), where each patch field stores the face-centre values
- surfaceFields, e.g. surfaceScalarField
    * internalField: for all internal face face-centre values
    * boundaryField: same as for volFields
- pointFields, e.g. pointScalarField [this one is different to the ones above]
    * internalField: values for all points (regardless of whether they are internal or not) are stored in this field
    * boundaryField: this does not store values, instead the point boundary condition classes are just used to enforce certain conditions on the internal field. The comments in the pointPatchField.H give more details:

```
Class                                                                                                                                                                                                        
    Foam::pointPatchField                                                                                                                                                                                    
                                                                                                                                                                                                             
Description                                                                                                                                                                                                  
    Abstract base class for point-mesh patch fields.                                                                                                                                                         
                                                                                                                                                                                                             
    The base-field does not store values as they are part of the                                                                                                                                             
    "internal field".  There are derived classes to store constraint values                                                                                                                                  
    e.g. fixedValuePointPatchField derived from the generic                                                                                                                                                  
    valuePointPatchField which ensures the values in the "internal field"                                                                                                                                    
    are reset to the fixed-values by applying the stored values.                                                                                                                                             
                                                                                                                                                                                                             
SourceFiles                                                                                                                                                                                                  
    pointPatchField.C                                                                                                                                                                                        
    pointPatchFieldNew.C     
```
I agree that is is a bit strange that pointFields are different but I guess they are used much less and so it evolved this way. I guess one reason is that points can be on more than one patch whereas faces cannot.

Consequently, accessing values on the boundary of pointFields (for writing or for reading) requires indexing into the internalField. In the current code, I use the `patch().meshPoints()`addressing to do this (i.e. give me the internal point index for each point on the patch); however, the pointPatchField class does have a couple of functions we could use instead for reading and writing to point boundaries: `patchInternalField()` and `addToInternalField(...)`. In this case, the code would become:

```
            const vectorField pif(displacementField.boundaryField()[patchID].patchInternalField());
            forAll(pointDisplacement_->boundaryField()[patchID], i)
            {
                for (unsigned int d = 0; d < dim; ++d)
                    buffer[i * dim + d] = pif[i][d];
            }
```
The `patchInternalField()` function should be called just once per patch because it has to construct the patch field and return a copy (too expensive to do for every point). As a result, I am not sure this is much clearer than using the meshPoints addressing directly. But I am open to your thoughts.

Reply:
Does this mean that our `Displacement::read` and `DisplacementDelta::read` methods are wrong/accidentally working?

Reply:
Hmnn... the `Displacement::read` code clearly works, so let me delve deeper:

OK, I found the following:

- The base class `pointPatchField` does **not** store any data
- Derived classes, like `fixedValuePointPatchField.H` derive from `pointPatchField` but they also derive from a class called `valuePointPatchField`
- The `valuePointPatchField`derives from `Field` and hence **does** store a **copy** of the patch point values, which are independent of the actual internal field point values
- When the `updateCoeffs()` function of the point boundary condition is called, it applies the patch copy of the point values (stored in `valuePointPatchField`) to the internal field, where they use the `setInInternalField` function.

So, conclusions:

- The current `Displacement::read` and `DisplacementDelta::read` methods are correct (as expected)
- The same method *should* work for the `Displacement::write` function (I will try it now)

I have learned something about point fields!

Reply:
I have learned **a lot** via this PR, thank you! :smile: 

Let's sleep on it and decide about the finality of the conclusion tomorrow. Other than that, I still need to make solids4foam run on my system... :see_no_evil:

Reply:
I agree.

For building solids4foam, I am looking at that now: I am building the nextRelease branch using GCC on my Ubuntu workstation. Once that works fine then I guess you should not have any issues!